File: 98.验证二叉搜索树.c
/*
 * @lc app=leetcode.cn id=98 lang=c
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
// bool isBST;
// struct TreeNode *last;

// void isVaild(struct TreeNode *root)
// {
//   if (root == NULL)
//     return;

//   if (root->left != NULL)
//     isVaild(root->left);

//   if (last != NULL && root->val <= last->val)
//   {
//     isBST = false;
//     return;
//   }
//   last = root;

//   if (root->right != NULL)
//     isVaild(root->right);
// }

bool isValidBST(struct TreeNode *root)
{
  // 法一：递归
  // isBST = true;
  // if (root != NULL)
  // {
  //   last = NULL;
  //   isVaild(root);
  // }
  // return isBST;
  // 法二：迭代
  struct TreeNode **stack, *node, *last;
  bool isBST;
  int top;
  top = 0;
  isBST = true;
  stack = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 10000);
  last = NULL;
  if (root != NULL)
  {
    stack[top++] = root;
    while (top > 0)
    {
      node = stack[--top];
      if (node != NULL)
      {
        if (node->right)
          stack[top++] = node->right;
        stack[top++] = node;
        stack[top++] = NULL;
        if (node->left)
          stack[top++] = node->left;
      }
      else
      {
        node = stack[--top];
        if (last != NULL && last->val >= node->val)
          isBST = false;
        last = node;
      }
    }
  }
  return isBST;
}
// @lc code=end

File: 98.验证二叉搜索树.js
/*
 * @lc app=leetcode.cn id=98 lang=javascript
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

let minVal;
const vaildBST = (root) => {
  if (!root) return true;
  console.log(minVal);
  let leftBST = isValidBST(root.left);

  if (minVal >= root.val) return false;
  else minVal = root.val;

  let rightBST = isValidBST(root.right);

  return leftBST && rightBST;
};
var isValidBST = function (root) {
  // minVal = -Math.pow(2, 31) - 1;
  // 递归
  // return vaildBST(root);
  // 迭代
  let stack = [],
    top = 0;
  let preVal = null;
  stack[top++] = root;
  while (top) {
    let node = stack[--top];
    if (!node) {
      node = stack[--top];
      if (preVal !== null && node.val <= preVal) return false;
      preVal = node.val;
    } else {
      if (node.right) stack[top++] = node.right;
      stack[top++] = node;
      stack[top++] = null;
      if (node.left) stack[top++] = node.left;
    }
  }
  return true;
};
// @lc code=end
